include/veriblock/storage/block_repository.hpp

```diff
@@ -0,0 +1,164 @@
+#include <memory>
+#include <vector>
+#include <veriblock/slice.hpp>
+
+namespace VeriBlock {
+
+namespace db {
+/**
+ * @class UnrecoverableError
+ * @brief Application is not able to recover from the error occurred in the
+ * storage.
+ *
+ * Example: database is corrupted and can not be opened.
+ *
+ * It is up to implementation to decide a full list of such error types.
+ */
+struct UnrecoverableError : public std::exception {};
+
+/**
+ * @class Error
+ * @brief Application in normal condition can recover from this type of error.
+ *
+ * Example: out of disk space, IO error.
+ *
+ * It is up to implementation to decide a full list of such error types.
+ */
+struct Error : public std::exception {};
+}  // namespace db
+
+/**
+ * @class BlockRepository
+ *
+ * @brief Represents a block tree stored on disk.
+ *
+ * Multiple blocks may be stored at the given height.
+ *
+ * @tparam Block Stored Block type
+ *
+ * @invariant any instance is in valid state - database is opened when valid
+ * instance of BlockRepository created. To close database, delete the instance.
+ */
+template <typename Block>
+struct BlockRepository {
+  //! stored block type
+  using stored_block_t = Block;
+  //! block has type
+  using hash_t = typename Block::hash_t;
+  //! block height type
+  using height_t = typename Block::height_t;
+
+  virtual ~BlockRepository() = default;
+
+  /**
+   * @class WriteBatch
+   * @brief Efficiently implements bulk write operation for BlockRepository.
+   *
+   * @invariant WriteBatch is always in valid state.
+   * @invariant WriteBatch does not modify on-disk storage after put/remove
+   * operations. It does, when \p BlockRepository::commit is executed on this
+   * batch.
+   */
+  struct WriteBatch {
+    virtual ~WriteBatch() = default;
+
+    /**
+     * Write a single block. If block with such hash exists, db will overwrite
+     * it.
+     * @param block to be written in a batch
+     * @return true if we overwrite existing block, false otherwise.
+     */
+    virtual bool put(const stored_block_t& block) = 0;
+
+    /**
+     * Remove a single block from storage identified by its hash.
+     * @param hash block hash
+     * @return true if removed, false if no such element found.
+     */
+    virtual bool removeByHash(const hash_t& hash) = 0;
+
+    /**
+     * Remove potentially many blocks at given height.
+     * @param height block height
+     * @return true if at least one block removed, false otherwise.
+     */
+    virtual bool removeByHeight(height_t height) = 0;
+
+    /**
+     * Clear batch from any modifying operations.
+     */
+    virtual void clear() = 0;
+  };
+
+  /**
+   * Load a block from disk in memory, by its hash.
+   * @param hash[in] block hash
+   * @param out[out] if non-null, block data will be written here. If null
+   * passed, out argument is ignored.
+   * @return true if block found, false otherwise.
+   */
+  virtual bool getByHash(const hash_t& hash, stored_block_t* out) = 0;
```

Comment:
this API looks soooo C-like. it's definitely asking for a pattern like Block get(Hash) and letting the compiler optimize the object creation/copy/move code.
if it is left like this, I'd at least add  a default =nullptr value for out or better yet, or provide 2 overloads, one that lacks out(could name it exists() while we're at it) and another where out is a reference

Reply:
re: lack of optional -- returning unique_ptr is an option but it wouldn't be as efficient as returning the block itself. Block get(Hash) that throws an exception if the block is not found might also make some sense. yes, my suggestions obviously cause the number of functions to increase, but that's what you'd need to get the compiler to do more static checking of your code

Reply:
although, we could always play the custom allocator/resource pool game with unique_ptr, so pretty interface != slow interface. but it does take work

Reply:
another option would be to return shared_ptr<const Block> and reuse the cached block instances

Reply:
This API is very similar to what RocksDB or LevelDB offer. I think we will stick to this API. It is low level and flexible enough to be usable. Also it is easy to implement and use.